module/scheduler_components/deterministic_lock_manager.cpp

```diff
@@ -196,6 +196,92 @@ DeterministicLockManager::ReleaseLocks(const Transaction& txn) {
   return ready_txns;
 }
 
+VerifyMasterResult
+DeterministicLockManager::VerifyMaster(const Transaction& txn) {
+  auto keys = ExtractKeys(txn);
+  if (keys.empty()) {
+    // None of the keys in this txn is in this partition
+    return VerifyMasterResult::Valid;
+  }
+
+  auto txn_master_metadata = txn.internal().master_metadata();
+  if (txn_master_metadata.empty()) { // allow this to not be set for testing
+    LOG(WARNING) << "Master metadata empty: txn id " << txn.internal().id();
+    return VerifyMasterResult::Valid;
+  }
+
+  // keeps track of counters that will need to be waited on
+  std::unordered_map<Key, int32_t> ahead_counters;
+  for (auto pair : keys) {
+    auto key = pair.first;
+
+    Record record;
+    bool found = storage_->Read(key, record);
+    if (found) {
+      CHECK(txn_master_metadata.contains(key))
+              << "Master metadata for key \"" << key << "\" is missing";
+
+      auto txn_metadata = txn_master_metadata.at(key);
+      auto stored_metadata = record.metadata;
+
+      if (txn_metadata.counter() > stored_metadata.counter) {
+        ahead_counters[key] += txn_metadata.counter() - stored_metadata.counter;
+      } else if (txn_metadata.counter() < stored_metadata.counter) {
+        return VerifyMasterResult::Aborted;
+      } else {
+        CHECK(txn_metadata.master() == stored_metadata.master)
+                << "Metadata is invalid, different masters with the same counter";
+      }
+    } else { // key not in storage
+      // TODO: verify if counter is 0?
+    }
+  }
+  if (!ahead_counters.empty()){
+    auto txn_id = txn.internal().id();
+    num_remasters_waited_[txn_id] = ahead_counters;
+
+    // add the txn to the overall queue for each key
+    for (auto key : ahead_counters) {
+      keys_waiting_remaster_[key.first].push_back(txn_id);
+    }
+    return VerifyMasterResult::Waiting;
+  } else {
+    return VerifyMasterResult::Valid;
+  }
+}
+
+list<TxnId> DeterministicLockManager::RemasterOccured(Key key) {
+  list<TxnId> unblocked;
+
+  if (keys_waiting_remaster_.count(key) == 0) {
+    return unblocked;
+  }
+
+  for (auto txn_id : keys_waiting_remaster_[key]) {
+    // the number of remasters this txn needs for each key
+    auto& waiting_map = num_remasters_waited_[txn_id];
+
+    waiting_map[key] -= 1;
+    if (waiting_map[key] == 0) { // no longer waiting on this key
+      waiting_map.erase(key);
+      if (waiting_map.empty()) { // no longer waiting on any key
+        unblocked.push_back(txn_id);
+        num_remasters_waited_.erase(txn_id);
+      }
+    }
+  }
```

Comment:
Good catch. I can add one more queue of txn ids in order, then try to remove the first and only continue to the next if the prior was unblocked.

Haven't thought of a way to do it without either an additional data structure or by making the updates require a scan

Reply:
I don't think you need any more queue or additional data structure. RemasterOccured is called for one key at a time so you can just do the check on the head of the queue and pop it until you no longer can. If you don't count to 0, you only need to maintain a static counter and compare it with the counter of the remaster transaction.

Reply:
I don't think you need any more queue or additional data structure. RemasterOccured is called for one key at a time so you can just do the check on the head of the queue and pop it until you no longer can. If you don't count to 0, you only need to maintain a static counter and compare it with the counter of the remaster transaction.

Reply:
In this case:
T1(A, B) and T2(A), both have correct counters on A. B counter is ahead by 1.
We need some way to block T2 right? Currently it would pass VerifyCounters and skip T1 in line